tests/unit_tests/instancetype_test.go

```diff
@@ -0,0 +1,178 @@
+)
+
+var _ = Describe("Common instance types unit tests", func() {
+
+	skipLabels := map[string]bool{
+		"os-type":    true,
+		"deprecated": true,
+		"version":    true,
+		"class":      true,
+	}
+
+	instanceTypeFunctionMap := map[string]func(string, string, instancetypev1beta1.VirtualMachineClusterInstancetype) error{
+		"cpu":                   checkCPU,
+		"memory":                checkMemory,
+		"hugepages":             checkHugepages,
+		"gpus":                  checkGPUs,
+		"numa":                  checkNuma,
+		"dedicatedCPUPlacement": checkDedicatedCPUPlacement,
+		"isolateEmulatorThread": checkIsolateEmulatorThread,
+		"vendor":                instancetypeCheckVendor,
+	}
+
+	preferenceFunctionMap := map[string]func(string, string, instancetypev1beta1.VirtualMachineClusterPreference) error{
+		"displayName": checkDisplayName,
+		"vendor":      preferenceCheckVendor,
+	}
+
+	Context("VirtualMachineClusterPreference", func() {
+
+		It("check version", func() {
+			for _, preference := range virtualMachineClusterPreferences {
+				Expect(preference.APIVersion).To(Equal(expectedVersion))
+			}
+		})
+
+		It("check if labels match resources", func() {
+			for _, preference := range virtualMachineClusterPreferences {
+				for key, value := range preference.Labels {
+					labelType := parseLabel(key)
+					if skipLabels[labelType] {
+						continue
+					}
+
+					if function, ok := preferenceFunctionMap[labelType]; ok {
+						Expect(function(value, labelType, preference)).ToNot(HaveOccurred())
+					} else {
+						Fail(fmt.Sprintf("Unknown label: %s", labelType))
+					}
+				}
+			}
+		})
+
+	})
+
+	Context("VirtualMachineClusterInstanceType", func() {
+
+		It("check version", func() {
+			for _, instanceType := range virtualMachineClusterInstancetypes {
+				Expect(instanceType.APIVersion).To(Equal(expectedVersion))
+			}
+		})
+
+		It("check if labels match resources", func() {
+			for _, instanceType := range virtualMachineClusterInstancetypes {
+				for key, value := range instanceType.Labels {
+					labelType := parseLabel(key)
+					if skipLabels[labelType] {
+						continue
+					}
+
+					if function, ok := instanceTypeFunctionMap[labelType]; ok {
+						Expect(function(value, labelType, instanceType)).ToNot(HaveOccurred())
+					} else {
+						Fail(fmt.Sprintf("Unknown label: %s", labelType))
+					}
+				}
+			}
+		})
+	})
+})
+
+func parseLabel(label string) string {
+	splitLabel := strings.Split(label, "/")
+	return splitLabel[len(splitLabel)-1]
+}
+
+func checkCPU(lableValue, labelName string, instanceType instancetypev1beta1.VirtualMachineClusterInstancetype) error {
+	expectedCPU, err := strconv.ParseUint(lableValue, 10, 32)
+	if err != nil {
+		return fmt.Errorf(instanceTypeErrorMessage, labelName, instanceType.Name)
+	}
+
+	if instanceType.Spec.CPU.Guest == uint32(expectedCPU) {
+		return nil
+	}
+
+	return fmt.Errorf(instanceTypeErrorMessage, labelName, instanceType.Name)
```

Comment:
Make it return nil by default and return err if err condition was met? Iow. swap L115 and L112 + adjust L111.